### ISSUE TITLE
perf(model): scope structural sharing for legacy Metal host prefix KV

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12856]
Parent: #12683

Coordinates with: #12528
Coordinates with: #12550
Coordinates with: #12831

## For / Problem / Today / Better because / Witness

- **For:** developers forking local Qwen3.8 agent sessions from one long shared prompt on Apple Silicon through the backend-nil legacy Metal path.
- **Problem:** `PrefixSnapshot.Clone` and `KVCache.Clone` copy every host `K`, `Kraw`, and `V` float slice before a reused session can append its divergent suffix. The captured OpenCode prompt is 34,127 tokens; adding the planned 1,024-token output reserve yields 35,151 rows. From the current F32 K/Kraw/V geometry, that prefix payload calculates to 6,910,967,808 bytes (about 6.436 GiB) per eager clone. This is a source-derived byte count, not a measured allocation, copy, resident-memory, or completed full native run.
- **Today:** `internal/model/kv.go` calls `Cache.Clone()` from both snapshot creation and snapshot cloning; `internal/model/kvcache.go` deep-copies all three planes. Prefix reuse can avoid repeated prefill while still paying a multi-gigabyte host-memory copy per subagent.
- **Better because:** an eventual backend-nil ownership design can share immutable prefix storage and isolate only divergent writes while keeping the original prefix exact.
- **Witness:** restore two sessions from one real prefix snapshot, append different suffixes through the actual legacy Metal session path, and prove the source snapshot plus both continuations remain exact. The witness must report copied prefix bytes and physical memory, not only clone-call latency.

## Dedupe and boundaries

- Closed #12528 binds synthetic `ctxmmu.KVPool` block IDs to radix/Metal page tables. Production `node.CloneKV()` still returns a flat deep-cloned `model.KVCache`; #12528 does not install a backend-nil `PrefixSnapshot` into the normal session or run divergent continuation.
- #12550 and #12831 own Qwen GDN recurrent tensor handles and device verification. They do not change the host softmax-attention `K/Kraw/V` planes used when `Backend == nil`.
- #12400 is the Vulkan device-KV copy-on-write leaf.
- #12191 and #12313 cover the general context-MMU pool and sharing telemetry. This ticket is the missing production integration contract for the legacy Metal host-KV owner.

## Scope status

**DISPATCHABLE.** The audit is complete and landed in
[`docs/notes/SCOPE-LEGACY-METAL-HOST-PREFIX-KV-STRUCTURAL-SHARING-2026-09-14.md`](https://github.com/anthony-chaudhary/fak/blob/main/docs/notes/SCOPE-LEGACY-METAL-HOST-PREFIX-KV-STRUCTURAL-SHARING-2026-09-14.md)
(commit `f633849bc`). See the scope-complete comment for the full inventory.

An append seam alone is insufficient evidence that existing `PagedKV.Fork` can replace flat
`KVCache.Clone`. The audit confirms it: non-test `internal/model` holds 42
`Cache.{K,Kraw,V}[...] = append(...)` sites, many on the *backend-nil legacy Metal* path
(`kv.go`, `metal_prefill*.go`, `metal_decode.go`, `qwen35_batch_decode_metal.go`,
`metal_qwen35_resident_decode.go`, `qwen35_prefill*.go`, `prefill_q4k.go`, `quant_forward.go`,
`kvcache_q8.go`), plus whole-slice reassignment in `evictSupported`/`Reserve`/`glm_dsa_session.go`/
`kv_tree_compaction.go`/`spanserialize.go`. Read-only attention consumers could tolerate an
indirection; writers cannot without copy-on-write or a typed fail-closed refusal.

Owned files for the implementation leaf (exactly three):

- `internal/model/kvcache.go` (layout owner; shared-prefix representation + COW guard before `Truncate`/`Reserve`/`evictSupported`)
- `internal/model/kv.go` (ownership funnel: `PrefixSnapshot`/`Clone`/`Restore`/`Close` `Backend == nil` branches)
- `internal/model/kv_cache_bytes.go` (+ `prefix_snapshot_bytes.go`; report shared bytes once)

Rollback: keep the current flat `[][]float32` layout as the default and gate sharing behind an
explicit capability (mirroring `FAK_PAGED_KV`), so reverting to `Clone()`'s deep copy restores
today's behavior with no consumer change.

Witness (green at HEAD, no GPU): `go test ./internal/model -run 'TestTokenLineage|TestSessionFromPrefix|TestPrefixSnapshot' -count=1`.

## Correctness witness

The software regression must exercise the actual session contract rather than an idle `Clone` microbenchmark:

1. Build a non-empty backend-nil Qwen session and capture a `PrefixSnapshot`.
2. Clone/restore two independently owned sessions.
3. Append distinct suffix tokens through the production legacy Metal path.
4. Prove the original prefix remains byte-exact in `K`, `Kraw`, `V`, positions, lineage, and logits.
5. Prove each branch matches an independently prefetched full prompt plus its suffix after divergence.
6. Prove close/eviction releases shared ownership exactly once.
7. Report prefix bytes copied before divergence and on first divergent append.

## Physical qualification remains separate

Software ownership completion does not qualify end-user performance. Preserve the existing 35,151-row/full-task native Qwen3.8 workflow as a separate physical promotion witness with exact artifact, source, prompt/task, context, engine/backend/no-fallback identity, peak footprint, paging, correctness, and task completion. This follow-on is not a new required criterion for the current prefix-reuse value proof under #12683.

## Non-goals

- GDN recurrent-handle COW.
- Vulkan device KV.
- A new generic KV-cache epic.
- An idle clone-only benchmark.
- Relaxing admission or memory-pressure guards.
- Claiming current `PagedKV` is already compatible with contiguous attention readers.

**Comment by anthony-chaudhary** [comment]
## Scope complete — dispatchable

Audit landed in `docs/notes/SCOPE-LEGACY-METAL-HOST-PREFIX-KV-STRUCTURAL-SHARING-2026-09-14.md` (commit `f633849bc`). The required inventory is done; the ticket is no longer `TRIAGE / SCOPE NEEDED`.

**Verdict: GO, scoped.** One bounded three-file slice now exists, with a fail-closed rollback that keeps the current flat `[][]float32` layout for every unsupported path. No sharing is implemented and no zero-copy/performance claim is made here.

### Byte figure verified

The 6,910,967,808-byte figure is exact, and it counts the 16 `full_attention` layers, not all 64:

```
35,151 rows × 16 KV-layers × 3 planes (K, Kraw, V) × 4 KVHeads × 256 HeadDim × 4 B = 6,910,967,808
```

`qwen38ContextPlanConfig()` (`internal/model/contextsize_test.go:88`) is 64 layers with `LayerTypes[i%4==3]="full_attention"` → 16 full + 48 linear. `planCacheGeometry` (`internal/model/cache_geometry.go:78`) charges all 64, so the budget and the payload disagree by 4×; the leaf must pick one deliberately.

### Boundary

`#12550`/`#12831` own the GDN/recurrent device COW; `#12528`/`#12191`/`#12313` own the ctxmmu/radixkv paged substrate; `#12400` owns Vulkan. None touches the legacy backend-nil host `K/Kraw/V` prefix. That lane is this issue's. (The `#12550` verifier defect at `qwen35_verify_panel_device.go:146` is repaired at HEAD; its residual reader-sync / reentrant-`Close` surface is scoped here, disjoint from its three files.)

### Owned files (future implementation leaf)

1. `internal/model/kvcache.go` — layout owner; shared-prefix representation + copy-on-write guard before `Truncate`, `Reserve`, `evictSupported`.
2. `internal/model/kv.go` — ownership funnel: `PrefixSnapshot` / `Clone` / `Restore` / `Close` `Backend == nil` branches.
3. `internal/model/kv_cache_bytes.go` (+ `prefix_snapshot_bytes.go`) — report shared bytes once.

### Why an append seam alone is insufficient

Non-test `internal/model` holds **42** `Cache.{K,Kraw,V}[...] = append(...)` sites, and they are not device-only. The legacy Metal path also appends host slices: `kv.go:1018,1024,1108`, `metal_prefill.go:195-197,281-295`, `metal_prefill_hybrid.go:588-590,880-882,1092-1094`, `metal_decode.go:162-163`, `qwen35_batch_decode_metal.go:97-99`, `metal_qwen35_resident_decode.go:566-593`, `qwen35_prefill*.go`, `prefill_q4k.go:173`, `quant_forward.go:722`, `kvcache_q8.go:275-303`. Whole-slice reassignment also lives in `evictSupported` / `Reserve` / `glm_dsa_session.go` / `kv_tree_compaction.go` / `spanserialize.go`. Read-only attention consumers could tolerate an indirection; the writers cannot without copy-on-write or a typed refusal.

### Rollback

Keep `[][]float32` as the default, gate sharing behind an explicit capability (mirroring `FAK_PAGED_KV`). Reverting the leaf to `Clone()`'s deep copy restores today's bytes with no consumer change.

### Witness (green at HEAD, no GPU)

```bash
go test ./internal/model -run 'TestTokenLineage|TestSessionFromPrefix|TestPrefixSnapshot' -count=1
```

Two independent checks confirmed this: a `cross-validator` verified every line/anchor claim against source (CLEAN, 4 line-precision fixes applied) and a `tester` re-ran the witness independently (GREEN; 6 test anchors matched).

Non-goals preserved: no GDN recurrent COW, no Vulkan device KV, no new epic, no clone-only benchmark, no admission relaxation, no physical/end-user performance claim. Physical promotion remains a separate witness on the 35,151-row workflow.
